src/Dfc.CourseDirectory.Web/Controllers/BulkUploadApprenticeshipsController.cs

```diff
@@ -131,126 +130,92 @@ public IActionResult Pending()
         public async Task<IActionResult> Index(IFormFile bulkUploadFile)
         {
             var sw = Stopwatch.StartNew();
-
             try
             {
-
-                int? UKPRN;
-                if (_session.GetInt32("UKPRN") != null)
-                    UKPRN = _session.GetInt32("UKPRN").Value;
-                else
-                    return RedirectToAction("Index", "Home", new { errmsg = "Please select a Provider." });
-
-                BulkUploadViewModel vm = new BulkUploadViewModel();
-
-
-
-                string errorMessage;
-
-                if (Validate.ValidateFile(bulkUploadFile, out errorMessage))
+                var UKPRN = _session.GetInt32("UKPRN");
+                if (UKPRN == null)
                 {
-                    int providerUKPRN = UKPRN.Value;
-                    string userId = User.FindFirst(ClaimTypes.NameIdentifier)?.Value;
-                    string bulkUploadFileNewName = string.Format(@"{0}-{1}", DateTime.Now.ToString("yyMMdd-HHmmss"),
-                         Path.GetFileName(bulkUploadFile.FileName));
-
-                    MemoryStream ms = new MemoryStream();
-                    bulkUploadFile.CopyTo(ms);
+                    return RedirectToAction("Index", "Home", new {errmsg = "Please select a Provider."});
+                }
+                if (!Validate.ValidateFile(bulkUploadFile, out var errorMessage))
+                {
+                    return View(new BulkUploadViewModel {errors = new[] {errorMessage}});
+                }
 
-                    if (!Validate.isBinaryStream(ms))
-                    {
-                        int csvLineCount = _apprenticeshipBulkUploadService.CountCsvLines(ms);
+                await using var ms = new MemoryStream(); // todo: await-using to dispose of stream. Needs test fix.
```

Comment:
No need to `Dispose` a `MemoryStream`

Reply:
And a todo left behind, oops

Reply:
According to https://stackoverflow.com/questions/4274590/memorystream-close-or-memorystream-dispose/4274769#4274769 you are correct. I wonder if we should dispose anyway given that it is disposable and that doesn't cause any issues. It would avoid IDE warnings and future discussions at least.

Reply:
I don't mind keeping the Dispose. It was more a comment on the TODO.

Reply:
Ah okay, magic. Todo is about to vanish

Reply:
Done